torch/csrc/api/include/torch/data/datasets/chunk.h

```diff
@@ -0,0 +1,432 @@
+          "Dataset needs to call reset() before calling get_batch().");
+    }
+    if (batch_size != batch_size_) {
+      std::string error =
+          "The requested batch size does not match with the initialized batch size.\n"
+          " The requested batch size is " + std::to_string(batch_size) +
+          ", while the data set is created with batch size equal to " +
+          std::to_string(batch_size_);
+
+      throw std::runtime_error(error);
+    }
+    return chunk_buffer_->get_batch(batch_size);
+  }
+
+  /// This will clear any internal state and starts the internal prefetching
+  /// mechanism for the chunk dataset.
+  virtual void reset() {
+    // free workers from previous reset if there is any.
+    free_workers();
+    preload_threads_.clear();
+
+    chunk_reader_.reset();
+
+    size_t chunks_to_load = chunk_reader_.get_chunk_count();
+    chunk_sampler_.reset(chunks_to_load);
+
+    // Creates a new chunk buffer each time we reset the dataset.
+    chunk_buffer_ = torch::make_unique<
+        BatchDataBuffer<UnwrappedBatchType, ExampleSamplerType>>(
+        chunks_to_load,
+        batch_size_,
+        example_sampler_,
+        ignore_empty_chunk_,
+        cache_size_);
+
+    // create new workers for this new epoch.
+    quit_worker_ = false;
+
+    for (size_t i = 0; i < preloader_count_; ++i) {
+      preload_threads_.emplace_back(
+          [this, i]() { this->preloader(i); });
+    }
+  }
+
+  /// size is not used for chunk dataset.
+  optional<size_t> size() const override {
+    return torch::nullopt;
+  }
+
+ private:
+  /// running on worker thread to preload chunk data.
+  void preloader(size_t id) {
+    while (!quit_worker_) {
+      size_t chunk_id;
+      try {
+        auto chunk_sampler_result = chunk_sampler_.next(1);
+        if (chunk_sampler_result.has_value()) {
+          chunk_id = chunk_sampler_result.value()[0];
+        } else {
+          break;
+        }
+        UnwrappedBatchType data = chunk_reader_.read_chunk(chunk_id);
+        if (data.empty()) {
+          if (!ignore_empty_chunk_) {
+            std::string error =
+                "Chunk with index " + std::to_string(chunk_id) + " is empty";
+            throw std::runtime_error(error);
+          } else {
+            // skip adding the current chunk data and move to the next.
+            chunk_buffer_->skip_chunk();
+          }
+        }
+        else {
+          chunk_buffer_->add_chunk_data(chunk_id, std::move(data));
+        }
+      } catch (...) {
+        chunk_buffer_->add_chunk_data(chunk_id, std::current_exception());
+      }
+    }
+  }
+
+  /// Block the current thread until the workers finish execution and exit.
+  void free_workers() {
+    if (!quit_worker_) {
+      quit_worker_ = true;
+      for (auto& worker_thread : preload_threads_) {
+        worker_thread.join();
+      }
+    }
+  }
+
+ private:
+  // chunk reader is responsible for reading chunk data
+  ChunkReader chunk_reader_;
```

Comment:
Again, the comment doesn't explain too much. I can guess that `ChunkReader` reads chunks, but there is zero documentation explaining what's the desired API of this type.